File: neetcode150/11_graphs/0684_redundant_connection.cpp
#include <bits/stdc++.h>
#include <chrono>
using namespace std;
/* LeetCode 684: Redundant Connection - Medium */
int par3[1001],rnk3[1001];
int find3(int x){while(par3[x]!=x){par3[x]=par3[par3[x]];x=par3[x];}return x;}
int main(){
    vector<pair<int,int>> edges={{1,2},{1,3},{2,3}};
    for(int i=0;i<=1000;i++){par3[i]=i;rnk3[i]=0;}
    auto s=chrono::high_resolution_clock::now();
    pair<int,int> ans={-1,-1};
    for(auto[a,b]:edges){int pa=find3(a),pb=find3(b);if(pa==pb){ans={a,b};break;}if(rnk3[pa]<rnk3[pb])swap(pa,pb);par3[pb]=pa;if(rnk3[pa]==rnk3[pb])rnk3[pa]++;}
    cout<<"var1: redundant=["<<ans.first<<","<<ans.second<<"]"<<endl;
    auto e=chrono::high_resolution_clock::now();
    cout<<"var1 time="<<chrono::duration_cast<chrono::nanoseconds>(e-s).count()/1000.0<<"us"<<endl;
}
// var1 mem = {} and time = {}

File: neetcode150/11_graphs/0684_redundant_connection.py
"""LeetCode 684: Redundant Connection - Medium"""
import time, tracemalloc
from typing import List

# Variation 1: Union Find - first edge that creates cycle
class Solution_v1:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        parent=list(range(len(edges)+1)); rank=[0]*(len(edges)+1)
        def find(x):
            while parent[x]!=x: parent[x]=parent[parent[x]]; x=parent[x]
            return x
        for a,b in edges:
            pa,pb=find(a),find(b)
            if pa==pb: return [a,b]
            if rank[pa]<rank[pb]: pa,pb=pb,pa
            parent[pb]=pa
            if rank[pa]==rank[pb]: rank[pa]+=1
        return []

# Variation 2: DFS - check if path exists before adding edge
class Solution_v2:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        from collections import defaultdict
        graph=defaultdict(set)
        def dfs(src, dst, visited):
            if src==dst: return True
            visited.add(src)
            for nb in graph[src]:
                if nb not in visited and dfs(nb,dst,visited): return True
            return False
        for a,b in edges:
            if dfs(a,b,set()): return [a,b]
            graph[a].add(b); graph[b].add(a)
        return []

if __name__ == "__main__":
    edges=[[1,2],[1,3],[2,3]]
    for Vi,Cls in enumerate([Solution_v1,Solution_v2],1):
        tracemalloc.start(); t0=time.perf_counter()
        r=Cls().findRedundantConnection(edges)
        t1=time.perf_counter(); mem=tracemalloc.get_traced_memory()[1]; tracemalloc.stop()
        print(f"var{Vi}: redundant={r}, mem={mem}bytes, time={(t1-t0)*1e6:.2f}us")
# var1 mem = {} and time = {}
# var2 mem = {} and time = {}
